### PR TITLE
Cache and print devices

### PR DESCRIPTION
**Pull request by julianKatz**
changes and printing the full list.

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/linkcache/cache.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+// it is ignored.
+func (d *linkCache) BrokenSymlink(symlink string) {
+	if entry, ok := d.devices[symlink]; ok {
+		entry.brokenSymlink = true
+		d.devices[symlink] = entry
+	}
+}
+
+func (d *linkCache) RemoveDevice(symlink string) {
+	delete(d.devices, symlink)
```
Comment (tonyzhc) [inline pkg/linkcache/cache.go:152]: Should we log here as well? Will that information be valuable?

File: pkg/linkcache/cache.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+	var errs []error
+	for _, entry := range entries {
+		if entry.IsDir() {
+			continue
+		}
+
+		// TODO(juliankatz): To have certainty this works for all edge cases, we
+		// need to test this with a manually partitioned disk.
+		if partitionNameRegex.MatchString(entry.Name()) {
```
Comment (tonyzhc) [inline pkg/linkcache/cache.go:83]: For my understanding, could you provide more context on why we skip partitions?

File: pkg/linkcache/cache.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+func (l *ListingCache) listAndUpdate() error {
+	visited := make(map[string]struct{})
+
+	entries, err := os.ReadDir(l.dir)
+	if err != nil {
+		return fmt.Errorf("failed to read directory %s: %w", l.dir, err)
+	}
+
+	var errs []error
```
Comment (tonyzhc) [inline pkg/linkcache/cache.go:75]: Maybe split out the rest of the function and make unit tests, so that you avoid dependency on os.ReadDir.

File: pkg/linkcache/cache.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+		d.devices[symlink] = entry
+	}
+}
+
+func (d *linkCache) RemoveDevice(symlink string) {
+	delete(d.devices, symlink)
+}
+
+func (d *linkCache) DeviceIDs() []string {
+	ids := make([]string, 0, len(d.devices))
```
Comment (tonyzhc) [inline pkg/linkcache/cache.go:156]: Can just be [maps.Keys](https://pkg.go.dev/golang.org/x/exp/maps#Keys)

File: pkg/linkcache/cache.go (hunk 2 above)
Comment (julianKatz) [reply pkg/linkcache/cache.go:83]: For the boot disks there will be something like 12 partitions.  Those end up being noise when we are answering the question of "which PV maps to which path on the VM".

File: pkg/linkcache/cache.go (hunk 4 above)
Comment (julianKatz) [reply pkg/linkcache/cache.go:156]: Done

File: pkg/linkcache/cache.go (hunk 3 above)
Comment (julianKatz) [reply pkg/linkcache/cache.go:96]: Turns out you can just make a mock filesystem.  Added unit test with that.

File: pkg/linkcache/cache.go (hunk 1 above)
Comment (julianKatz) [reply pkg/linkcache/cache.go]: Good call.  DOne.

File: pkg/linkcache/cache.go (hunk 2 above)
Comment (msau42) [reply pkg/linkcache/cache.go:104]: Add it as a comment in the code to make it easier for future readers

File: pkg/linkcache/cache.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+				klog.Warningf("Error listing and updating symlinks: %v", err)
+				continue
+			}
+
+			klog.Infof("periodic symlink cache read: %s", l.links.String())
+		}
+	}
+}
+
+func (l *ListingCache) listAndUpdate() error {
```
Comment (msau42) [inline pkg/linkcache/cache.go:88]: How do normal attach/detach flows show up with this? When we log, do we have a way to distinguish between normal attach/detach sequence vs when the device mapping unexpectantly changes post-mount?

I'm concerned about normal flows creating too much noise in the logs and drowning out the real issues.

File: pkg/linkcache/cache.go (hunk 2 above)
Comment (julianKatz) [reply pkg/linkcache/cache.go]: Moved to different model that doesn't have this problem

File: test/remote/instance.go [hunk 6]
```diff
@@ -235,7 +235,7 @@ func (i *InstanceInfo) CreateOrGetInstance(localSSDCount int) error {
 		}
 
 		if i.cfg.CloudtopHost {
```
Comment (julianKatz) [inline test/remote/instance.go:237]: This seems to no longer be necessary on cloudtop, and was actively causing my local test run to fail.

**Comment by sunnylovestiramisu** [comment]
Not under active development anymore.
